src/bit_vec.rs

```diff
@@ -68,72 +30,75 @@ impl<O: BitOrder, T: BitStore + ToByteSlice> Encode for BitVec<O, T> {
 
-			let mut aligned_vec: Vec<T> = vec![0u8.into(); required_bytes / mem::size_of::<T>()];
+/// Equivalent of `BitStore::MAX_BITS` on 32bit machine.
+const ARCH32BIT_BITSLICE_MAX_BITS: usize = 0x1fff_ffff;
 
-			unsafe {
-				let aligned_u8_ptr = aligned_vec.as_mut_ptr() as *mut u8;
-				for (i, v) in vec_u8.iter().enumerate() {
-					*aligned_u8_ptr.add(i) = *v;
-				}
+impl<O: BitOrder, T: BitStore + Decode> Decode for BitVec<O, T> {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		<Compact<u32>>::decode(input).and_then(move |Compact(bits)| {
+			// Otherwise it is impossible to store it on 32bit machine.
+			if bits as usize > ARCH32BIT_BITSLICE_MAX_BITS {
+				return Err("Attempt to decode a bitvec with too many bits".into());
 			}
-
-			let mut result = Self::from_vec(aligned_vec);
-			assert!(bits <= result.len());
-			unsafe { result.set_len(bits); }
+			let required_elements = required_elements::<T>(bits)? as usize;
+			let vec = decode_vec_with_len(input, required_elements)?;
+
+			let mut result = Self::try_from_vec(vec)
+				.map_err(|_| {
+					Error::from("UNEXPECTED ERROR: `bits` is less than
+					`ARCH32BIT_BITSLICE_MAX_BITS`; So BitVec must be able to handle the number of
+					segment needed for `bits` to be represented; qed")
+				})?;
+
+			assert!(bits as usize <= result.len());
+			result.truncate(bits as usize);
 			Ok(result)
 		})
 	}
 }
 
-impl<O: BitOrder, T: BitStore + ToByteSlice> Encode for BitBox<O, T> {
+impl<O: BitOrder, T: BitStore + Encode> Encode for BitBox<O, T> {
 	fn encode_to<W: Output>(&self, dest: &mut W) {
 		self.as_bitslice().encode_to(dest)
 	}
 }
 
-impl<O: BitOrder, T: BitStore + ToByteSlice> EncodeLike for BitBox<O, T> {}
+impl<O: BitOrder, T: BitStore + Encode> EncodeLike for BitBox<O, T> {}
 
-impl<O: BitOrder, T: BitStore + FromByteSlice> Decode for BitBox<O, T> {
+impl<O: BitOrder, T: BitStore + Decode> Decode for BitBox<O, T> {
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
 		Ok(Self::from_bitslice(BitVec::<O, T>::decode(input)?.as_bitslice()))
 	}
 }
 
-// Calculates bytes required to store given amount of `bits` as if they were stored in the array of `T`.
-fn required_bytes<T>(bits: usize) -> usize {
-	let element_bits = mem::size_of::<T>() * 8;
-	(bits + element_bits - 1) / element_bits * mem::size_of::<T>()
+/// Calculates the number of element `T` required to store given amount of `bits` as if they were
+/// stored in `BitVec<_, T>`
+///
+/// Returns an error if the number of bits + number of bits in element overflow u32 capacity.
+/// NOTE: this should never happen if `bits` is already checked to be less than
+/// `BitStore::MAX_BITS`.
+fn required_elements<T: BitStore>(bits: u32) -> Result<u32, Error> {
+	let element_bits = T::Mem::BITS as u32;
+	let error = Error::from("Attempt to decode bitvec with too many bits");
+	Ok((bits.checked_add(element_bits).ok_or_else(|| error)?  - 1) / element_bits)
```

Comment:
This checked_add is just extra safety but I'm ok to remove it with just a comment.